modyn/config/schema/pipeline/evaluation/strategy/_interval_strategy.py

```diff
@@ -0,0 +1,139 @@
+from functools import cached_property
+from typing import cast, get_args
+
+from modyn.config.schema.base_model import ModynBaseModel
+from modyn.const.regex import REGEX_TIME_UNIT
+from modyn.const.types import TimeUnit
+from pydantic import Field, field_validator, model_validator
+
+_EVAL_INTERVAL_BOUND_PATTERN = rf"[+-]?\s*((\d+\s*({REGEX_TIME_UNIT}))|(inf))"
+_EVAL_INTERVAL_PATTERN = (
+    rf"^(\[|\()\s*({_EVAL_INTERVAL_BOUND_PATTERN})\s*(,|;)\s*({_EVAL_INTERVAL_BOUND_PATTERN})\s*(\)|\])$"
+)
+
+
+class _IntervalEvalStrategyConfig(ModynBaseModel):
```

Comment:
Used for `PeriodicEvalStrategyConfig`. Can potentially replace the boundary selection of `OffsetEvalStrategy`.
However `OffsetEvalStrategy` currently serves two purposes:
- explicit: `the interval selection via a one-sided offset`
- implicit: trigger `time` at every training (with the training data batch start and end as interval center).

I'd like to keep `OffsetEvalStrategy` for now, but generally it could be expressed as a subclass of `_IntervalEvalStrategyConfig` that explicitly decodes the evaluation time at the point of trigger in its name.

Reply:
Could you clarify the implicit effect here? `OffsetEvalStrategy` can be used without ` after_training`, couldn't it? Or can it only be used after_training? I wonder -- based on our discussion yesterday -- how the `OffsetEvalStrategy` has an implicit effect on when we evaluate (at every training)

Reply:
the current version of `OffsetEvalStrategy` just takes `start_timestamp` and `end_timestamp` (the first and last timestamp of the training interval). Then it adds an offset to *ONE* of the intervals bounds of [start_timestamp, end_timestamp]. `_IntervalEvalStrategyConfig` is designed to do the same but is more flexible in its interface: 
- allows offsets at both ends
- allows `+0`, `-0` config syntax to refer to either `start_timestamp` or `end_timestamp`
- allows to specify if bounds are inclusive (through choice parenthesis).

This strategy can be executed after training (obviously we know `start_timestamp` and `end_timestamp` there. The `model=matrix` option does not make sense in this configuration though as we don't know the future `start/end_timestamps` just yet.

`after_training` we don't need to restrict the matrix option anymore.

I will add those restrictions and considerations directly into the config models, thanks for the hint!